packages/bloc/lib/src/bloc.dart

```diff
@@ -141,51 +153,53 @@ abstract class Bloc<Event, State> extends Stream<State> implements Sink<Event> {
   /// and return the new [state] in the form of a `Stream<State>`.
   Stream<State> mapEventToState(Event event);
 
-  /// Transforms the `Stream<State>` into a new `Stream<State>`.
-  /// By default [transformStates] returns the incoming `Stream<State>`.
-  /// You can override [transformStates] for advanced usage in order to
+  /// Transforms the `Stream<Transition>` into a new `Stream<Transition>`.
+  /// By default [transformTransitions] returns
+  /// the incoming `Stream<Transition>`.
+  /// You can override [transformTransitions] for advanced usage in order to
   /// manipulate the frequency and specificity at which `transitions`
   /// (state changes) occur.
   ///
-  /// For example, if you want to debounce outgoing [states]:
+  /// For example, if you want to debounce outgoing state changes:
   ///
   /// ```dart
   /// @override
-  /// Stream<State> transformStates(Stream<State> states) {
-  ///   return states.debounceTime(Duration(seconds: 1));
+  /// Stream<Transition<Event, State>> transformTransitions(
+  ///   Stream<Transition<Event, State>> transitions,
+  /// ) {
+  ///   return transitions.debounceTime(Duration(seconds: 1));
   /// }
   /// ```
-  Stream<State> transformStates(Stream<State> states) => states;
-
-  Stream<State> _buildStateStream() async* {
-    yield _state;
-    yield* _stateController.stream;
+  Stream<Transition<Event, State>> transformTransitions(
+    Stream<Transition<Event, State>> transitions,
+  ) {
+    return transitions;
   }
 
   void _bindEventsToStates() {
-    Event currentEvent;
-
-    transformStates(transformEvents(_eventController.stream, (event) {
-      currentEvent = event;
-      return mapEventToState(currentEvent).handleError(_handleError);
-    })).forEach(
-      (nextState) {
-        if (state == nextState || _stateController.isClosed) return;
-        final transition = Transition(
-          currentState: state,
-          event: currentEvent,
-          nextState: nextState,
-        );
-        try {
-          BlocSupervisor.delegate.onTransition(this, transition);
-          onTransition(transition);
-          _state = nextState;
-          _stateController.add(nextState);
-        } on dynamic catch (error) {
-          _handleError(error);
-        }
+    transformTransitions(transformEvents(
+      _eventController.stream,
+      (event) {
+        return mapEventToState(event).map((nextState) {
+          return Transition(
+            currentState: state,
+            event: event,
+            nextState: nextState,
+          );
+        }).skipWhile((transition) {
+          return state == transition.nextState || _stateController.isClosed;
+        }).handleError(_handleError);
       },
-    );
+    )).forEach((transition) {
```

Comment:
One other random Q: instead of using `forEach`, should this just `listen` to the Stream and store a `StreamSubscription _transitionSubscription` for cancellation in the `close` method?

Reply:
yeah good call 👍